index/scorch/merge.go

```diff
@@ -132,37 +132,54 @@ func (s *Scorch) planMergeAtSnapshot(ourSnapshot *IndexSnapshot) error {
 			if segSnapshot, ok := planSegment.(*SegmentSnapshot); ok {
 				oldMap[segSnapshot.id] = segSnapshot
 				if zapSeg, ok := segSnapshot.segment.(*zap.Segment); ok {
-					segmentsToMerge = append(segmentsToMerge, zapSeg)
+					if segSnapshot.LiveSize() == 0 {
```

Comment:
maybe i'm misunderstanding something, but this feels like an inconvenient way to drop segments that have live size of 0.  what i mean is, at this point in the code, we waited for the merger to plan to do something with a segment (combine it with some others) and now while walking through that list, we happen to notice that it's live size is 0.  we still don't do anything directly about it, but was pass it as nil in the old map, so that the introducer will end up dropping it when introducing this merge.

wouldn't it be easier if as a general rule, the introducer ALWAYS dropped segments with live size of 0.  not just for merges, like any operation it performs that copies over a root and creates a new one it also performs livesize=- dropping.  if we did that, this check here wouldn't be needed, as we would expect livesize=0 segments be dropped before the merger sees them.

Reply:
thats a very valid point here... we should add that logic in introducer too.
But since merge planner has a logic to pass in a task with empty segments, the logic for handling of any given tasks/segments has to be present here as well..